### PR TITLE
osd: handle removal of encrypted osd deployment

### PR DESCRIPTION
**Pull request by leseb**
This is handling a tricky scenario where the OSD deployment is manually
removed and the OSD never reconvers. This is unlikely to happen, but
still OSD should be able to run after that action. Essentially after a
manual deletion, we need to run the prepare job again to re-hydrate the
OSD information so that the OSD deployment can be deployed.
On encryption, it is a little bit tricky since ceph-volume list again
the main block won't return anything, so we need to target the encrypted
block to list.

There is another case this PR does not handle, which is the removal of
the OSD deployment and then the node is restarted. This means that the
encrypted container is not opened anymore. However, opening it requires
more work like writing the key on the filesystem (if not coming from the
Kubernete secret, eg,. KMS vault) and then run luksOpen. This is an
extreme corner case probably not worth worrying about for now.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
As far as code, the logic here is a little hard to follow, but I trust that it's the right thing to do here. I am having a little bit of trouble trying to understand the underlying case and mode of failure.

Notably, I'm not quite following the below text from the commit message:
> ... so that the OSD deployment can be deployed.

File: pkg/daemon/ceph/osd/volume.go [hunk 1]
```diff
@@ -171,22 +171,22 @@ func (a *OsdAgent) configureCVDevices(context *clusterd.Context, devices *Device
...
-			return osds, nil
-		}
-
-		// List existing OSD(s) configured with ceph-volume lvm mode
-		lvmOsds, err = GetCephVolumeLVMOSDs(context, a.clusterInfo, a.clusterInfo.FSID, lvPath, false, false)
-		if err != nil {
-			logger.Infof("failed to get devices already provisioned by ceph-volume. %v", err)
-		}
-		osds = append(osds, lvmOsds...)
+		} else {
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/volume.go:174]: Why we need this to be`else`? Versus just continuing on after the `if`?

**Comment by BlaineEXE** [comment]
Also, a bit of a complaint: what is the good of ceph-volume if we have to do utterly so much work to work around it and have so many special cases?

**Comment by leseb** [comment]
> As far as code, the logic here is a little hard to follow, but I trust that it's the right thing to do here. I am having a little bit of trouble trying to understand the underlying case and mode of failure.
> 
> Notably, I'm not quite following the below text from the commit message:
> 
> > ... so that the OSD deployment can be deployed.

Let's say `deploy/rook-ceph-osd-0` is removed manually. The operator reconciles the CephCluster on deletion events for the OSD deployment. At this stage, the operator must run the prepare job again to re-initialize the OSD information in order to run the OSD deployment again.
That's what  I meant by this sentence. Hope if clarifies.

**Comment by leseb** [comment]
> Also, a bit of a complaint: what is the good of ceph-volume if we have to do utterly so much work to work around it and have so many special cases?

The problem is that the raw/pvc case in c-v has been left behind a little bit and the encryption piece is only used by Rook as far as I can tell. Also, it's hard for c-v to know all the details we do in Rook.

File: pkg/daemon/ceph/osd/volume.go (hunk 1 above)
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go:174]: I just felt that the `else` really breaks the statements, it feels easier to read and parse that large code block.

File: pkg/operator/ceph/cluster/osd/create.go [hunk 2]
```diff
@@ -190,6 +190,13 @@ func (c *Cluster) startProvisioningOverPVCs(config *provisionConfig, errs *provi
 		}
 		osdProps.storeConfig.DeviceClass = volume.CrushDeviceClass
 
+		// Skip OSD prepare if deployment already exists for the PVC
+		// Also skip the encryption work part to avoid overriding the existing encryption key
+		if existingDeployments.Has(dataSource.ClaimName) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/create.go:195]: This is the same check that is seen below on line 242, right? Seems like we can delete the check on line 242 since this covers it now, just for encrypted OSDs in addition?

**Review by travisn — APPROVED** [review]
Have we seen this issue happen in production? Seems like it's only a test scenario where someone intentionally deleted an osd deployment and we need to go create it again.
A couple small suggestions, and approving since i'll be out this week...

File: pkg/operator/ceph/cluster/osd/create.go (hunk 2 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/create.go:195]: Weird and good catch, my goal was to move up this check.

**Comment by leseb** [comment]
Comments have been addressed.
